imgui/src/drag_drop.rs

```diff
@@ -0,0 +1,534 @@
+#[non_exhaustive]
+pub struct DragDropPayloadEmpty {
+    /// Set when [`accept_payload_empty`](Self::accept_payload_empty) was called
+    /// and mouse has been hovering the target item.
+    pub preview: bool,
+
+    /// Set when [`accept_payload_empty`](Self::accept_payload_empty) was
+    /// called and mouse button is released over the target item.
+    pub delivery: bool,
+}
+
+/// A DragDropPayload with status information and some POD, or plain old data,
+/// in it.
+#[derive(Debug, Copy, Clone)]
+#[non_exhaustive]
+pub struct DragDropPayloadPod<T: 'static + Copy> {
+    /// The kind data which was requested.
+    pub data: T,
+
+    /// Set when [`accept_payload`](Self::accept_payload) was called
+    /// and mouse has been hovering the target item.
+    pub preview: bool,
+
+    /// Set when [`accept_payload`](Self::accept_payload) was
+    /// called and mouse button is released over the target item.
+    pub delivery: bool,
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct DragDropPayload {
+    /// Data which is copied and owned by ImGui. If you have accepted the payload, you can
+    /// take ownership of the data; otherwise, view it immutably. Interacting with `data` is
+    /// very unsafe.
+    pub data: *const ffi::c_void,
+
+    /// The size of the data in bytes.
+    pub size: usize,
+
+    /// Set when [`accept_payload_unchecked`](Self::accept_payload_unchecked) was called
+    /// and mouse has been hovering the target item.
+    pub preview: bool,
+
+    /// Set when [`accept_payload_unchecked`](Self::accept_payload_unchecked) was
+    /// called and mouse button is released over the target item. If this is set to false, then you
+    /// set DragDropFlags::ACCEPT_BEFORE_DELIVERY and shouldn't mutate `data`.
+    pub delivery: bool,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+struct TypedPayload<T> {
+    header: TypedPayloadHeader,
+    data: T,
+}
+
+/// We have this struct separately to easily read it off in unsafe code.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+#[repr(C)]
+struct TypedPayloadHeader {
+    type_id: any::TypeId,
+    #[cfg(debug_assertions)]
+    type_name: &'static str,
+}
+
+impl TypedPayloadHeader {
+    #[cfg(debug_assertions)]
+    fn new<T: 'static>() -> Self {
+        Self {
+            type_id: any::TypeId::of::<T>(),
+            type_name: any::type_name::<T>(),
+        }
+    }
+
+    #[cfg(not(debug_assertions))]
+    fn new<T: 'static>() -> Self {
+        Self {
+            type_id: any::TypeId::of::<T>(),
+        }
+    }
+}
+
+impl<T: Copy + 'static> TypedPayload<T> {
+    /// Creates a new typed payload which contains this data.
+    pub fn new(data: T) -> Self {
+        Self {
+            header: TypedPayloadHeader::new::<T>(),
+            data,
+        }
+    }
+}
+
+/// Indicates that an incorrect payload type was received. It is opaque,
+/// but you can view questionably useful debug information with Debug formatting.
```

Comment:
> questionably useful

Yeah I mean, having the type name in debug builds would helps a lot. I've had to debug "what the fuck is in this `Box<dyn Any>`"-issues before and it was... not fun.

That said if you don't think it's useful, feel free to push back on review comments I make.

Reply:
ah no sorry, this comment predates adding the typename -- I hadn't thought of that idea at the time. Will remove -- I was previously just dumping out the typeid which is definitely "questionably" useful at best since it's not like a user could tell what that was. Now it's definitely useful